### PR TITLE
feat(policies): add WBCPolicy provider for GR00T Whole-Body-Control (ONNX)

### PR DESCRIPTION
**Pull request by cagataycali**
## Problem

No lightweight locomotion policy provider exists for humanoid robots (Unitree G1). The existing `groot` provider targets the full SONIC stack (Isaac-Lab dependent, training-grade). Users wanting deploy-grade whole-body-control have no path within strands-robots.

## Change

Add `WBCPolicy` - a lightweight ONNX-only policy provider implementing the decoupled Whole-Body-Control architecture from NVlabs/GR00T-WholeBodyControl.

**Architecture:** RL lower-body controller (12 leg + 3 waist joints = 15-dim action) with the upper 14 arm joints free for composition with VLA/IK/scripted policies.

**Key properties:**
- `requires_images=False` (state-only, high throughput)
- Auto-switches between Balance and Walk ONNX models based on velocity command magnitude
- 6-step observation history (86-dim per step, 516-dim flattened ONNX input)
- No PyTorch dependency: onnxruntime-only (~8MB per model)
- Models auto-downloaded from HuggingFace `nvidia/GR00T-WholeBodyControl` on first use
- Runs 5x realtime on CPU (tested headless on aarch64)
- Refuses to construct silently on missing models: raises `RuntimeError` unless `allow_missing_models=True` is explicitly passed

**Usage:**
```python
from strands_robots.policies import create_policy

policy = create_policy("wbc", checkpoint="nvidia/GR00T-WholeBodyControl")
policy.set_robot_state_keys(WBC_JOINT_NAMES)
actions = policy.get_actions_sync(obs, "", target_velocity=[0.5, 0, 0])
```

Well-known kwargs: `target_velocity=[vx, vy, wz]`, `target_height=0.74`, `target_rpy=[r, p, y]`

## Files

| File | Purpose |
|------|--------|
| `strands_robots/policies/wbc/__init__.py` | Package init, exports |
| `strands_robots/policies/wbc/wbc_policy.py` | WBCPolicy implementation |
| `tests/policies/wbc/test_wbc_policy.py` | Unit tests (mocked ONNX, obs encoding, factory integration) |
| `tests/policies/wbc/test_wbc_integration.py` | Integration tests (real ONNX, skipped without models) |
| `strands_robots/registry/policies.json` | Provider registration (shorthands: `wbc`, `groot_wbc`) |
| `pyproject.toml` | Added `onnxruntime` to mypy ignore list |

## Tests

Unit tests cover observation encoding (command scaling, gravity projection, joint error, history), action shape (15-dim), no-NaN guarantee, Balance vs Walk auto-switch, history buffer semantics, factory integration, and the fatal-on-missing-models contract (raises `RuntimeError`) plus the explicit `allow_missing_models=True` opt-in path.

## Follow-up (separate PRs)

- Integration tests with actual ONNX models in MuJoCo sim
- Asset download helper for G1 full-body XML (29-DOF scene)
- Example script + docs page
- CompositePolicy wrapper for WBC + upper-body VLA stacking

## License

Code: Apache 2.0 (matches this repo). ONNX weights fetched from HF at runtime under NVIDIA Open Model License - users opt-in by calling `create_policy("wbc")`; no weights bundled in this repo.

---

## Review changelog

**Round 1** (54caf88) - safety hardening from review feedback:
- `_load_models` now raises `RuntimeError` when both Balance and Walk ONNX sessions fail to load, instead of warn-and-continue (AGENTS.md conventions #5 / #6). Refuses to silently command default-pose targets on a bipedal humanoid.
- Degraded offline-testing path gated behind explicit `allow_missing_models=True` kwarg.
- Tightened the bare `except Exception` at the HuggingFace download site to `except (OSError, ImportError, ValueError, ConnectionError)`.
- Added PD torque-control path and e2e tests.

**Round 2** (6a0a5dd) - CI fix:
- 5 construction/factory unit tests failed in minimal CI because `onnxruntime` (optional, deploy-grade) is not installed there, so the `import onnxruntime` guard raised `ImportError` before the missing-model/factory logic under test could run.
- Added an absent-only autouse fixture stubbing `onnxruntime` in `sys.modules` when the real wheel is unavailable, mirroring the torch-mock doctrine in `tests/conftest.py`. When the wheel IS installed the genuine package is used. All 36 unit tests now pass (was 31 passed / 5 failed).

### CODE REVIEW

File: strands_robots/policies/wbc/wbc_policy.py [hunk 1]
```diff
@@ -0,0 +1,426 @@
...
+            logger.info("Loaded walk ONNX: %s", walk_path)
+
+        if self._balance_session is None and self._walk_session is None:
+            logger.warning(
+                "No ONNX models loaded from checkpoint '%s'. "
+                "Policy will return zero actions until models are available. "
+                "Download with: huggingface-cli download %s",
+                self._checkpoint,
+                self._checkpoint,
+            )
```
Comment (yinsong1986) [inline strands_robots/policies/wbc/wbc_policy.py]: [MUST FIX] When both ONNX downloads fail, `_load_models` only `logger.warning(...)` and returns. `_select_session` then returns `None` (both `or` chains fall through), `_run_inference` returns `np.zeros(_ACTION_DIM)` (line 331), and `get_actions` produces `action = 0 * 0.25 + self._default_angles` — i.e. silently commands the humanoid to its hard-coded standing pose (knees 0.3 rad, ankles -0.2 rad, etc.).

For a balance/walk policy on a 1.3m bipedal humanoid this is a real safety hazard: a user who misspells `checkpoint=`, lacks HF connectivity, or hits the `except Exception` swallow at line 241–242 gets a constructed `WBCPolicy` that *appears* to work and quietly issues default-pose targets every tick.

This violates two explicit AGENTS.md key conventions:
- #5 "Raise on fatal errors — never warn-and-continue if the system will behave unexpectedly"
- #6 "No silent defaults on error — returning zero-valued actions on failure is forbidden"

Resolution: in `_load_models`, raise `RuntimeError` (or a typed `PolicyLoadError`) when both `_balance_session` and `_walk_session` are `None`, instead of warning. If the user genuinely wants a no-op for offline testing, expose that behind an explicit `allow_missing_models=True` kwarg the way other providers do for opt-in degraded modes. Tightening the bare `except Exception` at line 241 to `except (OSError, ImportError, ValueError)` (or whatever `huggingface_hub` actually raises) would also be worth doing in the same pass so genuine bugs aren't masked at debug-log level.

**Review by yinsong1986 — COMMENTED** [review]
## Summary

Adds a new `WBCPolicy` provider implementing GR00T Whole-Body-Control (decoupled architecture) as an ONNX-only locomotion policy for humanoids (Unitree G1 class). 15-dim lower-body action (12 leg + 3 waist), 86-dim observation x 6-step history flattened to a 516-dim ONNX input, with auto-switching between Balance and Walk checkpoints on the magnitude of the velocity command. New registry entries (`wbc`, `groot_wbc`), no PyTorch dep (onnxruntime only), 34 unit tests with mocked sessions.

## What's good

- Provider follows the `Policy` base contract (`requires_images=False`, well-known kwargs `target_velocity` / `target_height` / `target_rpy`, `set_robot_state_keys`).
- Optional dep guarded by `require_optional`-style import error in `_load_models` and added to mypy ignore list in `pyproject.toml`.
- ONNX weights are not bundled; pulled from HF on first use under NVIDIA Open Model License — clean separation from the Apache-2.0 code in this repo.
- 34 tests are well-shaped (constants, quaternion math, mocked-session shape assertions, factory integration).

## Verification suggestions

- Run the full suite with `hatch run test` to confirm the 34 new tests pass alongside the existing 3496.
- Spot-check the registry round-trip: `python -c "from strands_robots.policies import create_policy, list_providers; print('wbc' in list_providers())"`.
- For an integration smoke test, `huggingface-cli download nvidia/GR00T-WholeBodyControl` then construct `WBCPolicy(checkpoint=...)` and confirm both ONNX files load (or, after addressing the inline finding, that a missing checkpoint raises rather than warns).

**Comment by cagataycali** [comment]
## Addressed MUST FIX + End-to-End Verification

Pushed commit addressing the review feedback.

### MUST FIX resolved

`_load_models` now **raises `RuntimeError`** when both ONNX sessions fail to load, instead of warning and silently commanding default-pose targets. Users who genuinely want a no-op for offline testing pass `allow_missing_models=True` (explicit opt-in to degraded mode).

Also tightened the bare `except Exception` in `_resolve_onnx_path` to `(OSError, ImportError, ValueError, ConnectionError)` so genuine bugs surface.

### CodeQL fix

`_OBS_INPUT_DIM` is now used in an assertion inside `get_actions` to validate the flattened observation history shape at runtime.

### End-to-End Integration (verified on aarch64)

Added `compute_torques()` helper method (PD controller with verified G1 kp/kd gains from upstream config) and 5 integration tests running **real ONNX models** in MuJoCo:

| Scenario | Duration | Min Pelvis Z | Final Pelvis Z | Forward Distance | Status |
|----------|----------|-------------|----------------|-----------------|--------|
| Balance  | 5s       | 0.743m      | 0.747m         | -0.076m         | PASS   |
| Walk fwd | 5s       | 0.740m      | 0.747m         | 2.069m          | PASS   |
| Turn     | 5s       | 0.740m      | 0.741m         | -0.063m         | PASS   |

**41/41 WBC tests pass** (36 unit + 5 integration). Robot stays balanced, walks forward at commanded velocity, turns in place without falling.

Integration tests are gated behind `pytest -m wbc` marker and skip gracefully when ONNX models aren't available locally.

### Key insight from e2e testing

The `g1_gear_wbc.xml` uses **torque actuators** (not position). WBCPolicy outputs target joint positions, which must be converted to torques via PD control. The new `compute_torques()` helper exposes this cleanly:

```python
target_pos = policy.get_actions_sync(obs, "", target_velocity=[0.5, 0, 0])
torques = policy.compute_torques(target_pos_array, current_pos, current_vel)
data.ctrl[:15] = torques
```

This is the correct integration pattern for torque-actuated models, matching the upstream NVlabs reference implementation.

File: strands_robots/policies/wbc/wbc_policy.py (hunk 1 above)
Comment (cagataycali) [reply strands_robots/policies/wbc/wbc_policy.py]: Addressed in 54caf88. `_load_models` now raises `RuntimeError` when both Balance and Walk sessions are `None`, refusing to construct a policy that would silently command default-pose targets on a bipedal humanoid (AGENTS.md conventions #5 and #6). The opt-in degraded path for offline testing is gated behind an explicit `allow_missing_models=True` kwarg, matching how other providers expose degraded modes. The bare `except Exception` at the HF-download site was tightened to `except (OSError, ImportError, ValueError, ConnectionError)` so genuine bugs are no longer masked at debug-log level.

Pinned regression tests: `test_no_models_raises_by_default` and `test_create_wbc_raises_without_models` assert the `RuntimeError`; `test_no_models_returns_default_actions_when_allowed` covers the explicit opt-in path.
